keps/sig-api-machinery/2558-publish-version-openapi/README.md

```diff
@@ -0,0 +1,421 @@
+for each of these feature gates in [`kube_features.go`] lists
+the prerelease status of the feature gate and the minimum Kubernetes
+version for the particular prerelease. These comments do not follow
+a pattern and are not parseable.
+
+Currently, when a new field is added, there is no way to map and document
+the field with the feature gate that controls it. Some fields mention
+the feature gates in a comment but these comments are not parseable either.
+
+**Field version**
+
+Currently, when a new field is added to a stable API version, there
+is no way to expose the version of the new field in OpenAPI.
+One way to identify alpha API fields is by comparing the OpenAPI schema
+generated from an apiserver with all APIs enabled and the one generated
+from an apiserver with only stable and beta APIs enabled.
+However, this process is cumbersome and not always feasible because not all
+fields are gated at the API group level.
+
+Since the OpenAPI definition is the source of truth for API documentation,
+this KEP proposes a way to publish versioning and feature gate
+information via OpenAPI.
+
+As a follow-up, this information can also be used to:
+- update the Kubernetes API reference documentation
+- surface [server-side warnings] in client-go and kubectl when a deprecated
+field or alpha field is being used.
+- surface warnings next to the field name in IDE when a deprecated
+field or alpha field is being used.
+
+### Goals
+
+- Publish feature gate associated with an API field for built-in API types
+in OpenAPI to document the mapping of API fields with their feature gates.
+- Document the current prerelease status of an API field by publishing
+the prerelease status of the field and the minimum Kubernetes version for
+the prerelease in OpenAPI.
+- Validate that the feature gate mentioned in the API field comment
+is a valid feature gate.
+
+### Non-Goals
+
+- Support publishing versioning information for built-in `Kind`s.
+This may be added in the future but is not in scope in the alpha status of this KEP.
+- Support publishing field-level versioning for custom resources in OpenAPI.
+- Publish the default status of a feature gate (enabled/disabled by default)
+in OpenAPI.
+- Allow parsing comment tags on feature gates in [`kube_features.go`].
+- Update the Kubernetes API reference docs to include the field-level
+versioning information (out of scope of this KEP, can be done in a follow up).
+
+## Proposal
+
+This KEP proposes changes to the kube-openapi generator to add a new
+vendor extension field `x-kubernetes-api-lifecycle` in OpenAPI.
+This extension specifies the prerelease status of an API field,
+the minimum Kubernetes version for the prerelease and the
+feature gate associated with the field.
+
+This information is derived from comment tag on API fields with the format:
+
+```
+// +lifecycle:kubernetes:minVersion=<k8s-version>,status=<prelease-status>,featureGate=<featuregate-name>
```

Comment:
this is currently going to be difficult to implement in controller-gen -- controller-gen treats the "name" part of a marker as static identifying information (so in this case, it'd treat the name as `+lifecycle:kubernetes`).  Theoretically, we could overhaul the way controller-gen interprets names to allow for name hierarchies or something, but it'd be *way* more straightforward to just do something like

```go
// +lifecycle:component=kubernetes,minVersion=<ver>,status=<status>,featureGate=<featuregate>
```

For example

```go
// +lifecycle:component=kubernetes,minVersion=v1.12,status=alpha,featureGate=Blah
// +lifecycle:component=my-company-thing,minVersion=v3.9,status=beta,featureGate=Other
```

Reply:
lgtm

Reply:
More specificially, the form in the KEP is "fine" if we only want a hardcoded list, but if the CRD form supports arbitrary component names, that's where we're gonna need the rework.

Reply:
@DirectXMan12 Thank you! Created https://github.com/kubernetes/enhancements/pull/2752 to fix this